File: electron/database/timeMemoRepository.ts
import type Database from "better-sqlite3";
import type { TimeMemo } from "../types";

interface TimeMemoRow {
  id: string;
  date: string;
  hour: number;
  content: string;
  created_at: string;
  updated_at: string;
}

function rowToNode(row: TimeMemoRow): TimeMemo {
  return {
    id: row.id,
    date: row.date,
    hour: row.hour,
    content: row.content,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function createTimeMemoRepository(db: Database.Database) {
  const stmts = {
    fetchByDate: db.prepare(
      `SELECT * FROM time_memos WHERE date = ? ORDER BY hour`,
    ),
    upsert: db.prepare(`
      INSERT INTO time_memos (id, date, hour, content, created_at, updated_at)
      VALUES (@id, @date, @hour, @content, datetime('now'), datetime('now'))
      ON CONFLICT(date, hour) DO UPDATE SET
        content = @content, updated_at = datetime('now')
    `),
    delete: db.prepare(`DELETE FROM time_memos WHERE id = ?`),
    fetchById: db.prepare(`SELECT * FROM time_memos WHERE id = ?`),
    fetchByDateHour: db.prepare(
      `SELECT * FROM time_memos WHERE date = ? AND hour = ?`,
    ),
  };

  return {
    fetchByDate(date: string): TimeMemo[] {
      return (stmts.fetchByDate.all(date) as TimeMemoRow[]).map(rowToNode);
    },

    upsert(id: string, date: string, hour: number, content: string): TimeMemo {
      stmts.upsert.run({ id, date, hour, content });
      const row = stmts.fetchByDateHour.get(date, hour) as TimeMemoRow;
      return rowToNode(row);
    },

    delete(id: string): void {
      stmts.delete.run(id);
    },
  };
}

export type TimeMemoRepository = ReturnType<typeof createTimeMemoRepository>;

File: src-tauri/src/db/time_memo_repository.rs
use rusqlite::{params, Connection};
use serde::{Deserialize, Serialize};

use super::helpers;
use super::row_converter::{query_all, query_one, FromRow};

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct TimeMemo {
    pub id: String,
    pub date: String,
    pub hour: i64,
    pub content: String,
    pub created_at: String,
    pub updated_at: String,
}

impl FromRow for TimeMemo {
    fn from_row(row: &rusqlite::Row) -> rusqlite::Result<Self> {
        Ok(TimeMemo {
            id: row.get("id")?,
            date: row.get("date")?,
            hour: row.get("hour")?,
            content: row.get("content")?,
            created_at: row.get("created_at")?,
            updated_at: row.get("updated_at")?,
        })
    }
}

pub fn fetch_by_date(conn: &Connection, date: &str) -> rusqlite::Result<Vec<TimeMemo>> {
    query_all(
        conn,
        "SELECT * FROM time_memos WHERE date = ?1 ORDER BY hour ASC",
        [date],
    )
}

pub fn upsert(
    conn: &Connection,
    id: &str,
    date: &str,
    hour: i64,
    content: &str,
) -> rusqlite::Result<TimeMemo> {
    let now = helpers::now();
    conn.execute(
        "INSERT INTO time_memos (id, date, hour, content, created_at, updated_at) \
         VALUES (?1, ?2, ?3, ?4, ?5, ?6) \
         ON CONFLICT(date, hour) DO UPDATE SET \
         content = excluded.content, updated_at = excluded.updated_at",
        params![id, date, hour, content, &now, &now],
    )?;

    query_one(
        conn,
        "SELECT * FROM time_memos WHERE date = ?1 AND hour = ?2",
        params![date, hour],
    )
}

pub fn delete(conn: &Connection, id: &str) -> rusqlite::Result<()> {
    conn.execute("DELETE FROM time_memos WHERE id = ?1", [id])?;
    Ok(())
}
